### PR TITLE
Ensure that inference dummy_forward does not try to match on a cudagraph when running eager

### PR DESCRIPTION
**Pull request by mathemakitten**
# What does this PR do ?
Fix for #3789.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py
Comment (sidsingh-nvidia) [inline megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py]: can we make this @torch inference mode?

File: megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py [hunk 1]
```diff
@@ -145,7 +145,23 @@ def dummy_forward(self):
...
+        # If the controller is running the graphed dummy_forward, it will run
+        # `_dynamic_step_forward_logits` instead. Setting this to False ensures that we
+        # will not try to match on a cudagraph when this is running eager.
+        self.inference_context._using_cuda_graph_this_step = False
+
+        # Pass inference_context so that transformer & mamba layers use the inference-mode
+        # cudagraph check (which gates on using_cuda_graph_this_step()) instead of
+        # the training-mode check (which unconditionally replays captured graphs
+        # when inference_context is not in kwargs).
+        return self.model(
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py]: I'm a bit scared that this will lead to segfaults. If we pass the inference context, we will attempt to write into the kv-cache/mamba-cache. But we are not setting up the required pointers properly.

Comment (mathemakitten) [reply megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py]: Removed this in favour of gating on inference_mode instead
